common/src/main/java/io/pravega/common/util/btree/sets/BTreeSet.java

```diff
@@ -0,0 +1,414 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util.btree.sets;
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Preconditions;
+import io.pravega.common.TimeoutTimer;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.ArrayView;
+import io.pravega.common.util.AsyncIterator;
+import io.pravega.common.util.ByteArrayComparator;
+import java.time.Duration;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+/**
+ * A B+Tree-backed Set. Stores all items in a B+Tree Structure using a {@link ByteArrayComparator} for ordering them.
+ *
+ * NOTE: This component is in {@link Beta}. There are no guarantees about data or API compatibility with future versions.
+ * Any component that is directly dependent on this one should either be in {@link Beta} as well.
+ */
+@NotThreadSafe
+@Beta
+@Slf4j
+public class BTreeSet {
```

Comment:
If I haven't already, we should file a follow-up issue to unify as much as we can between the existing BTreeIndex and this BTreeSet.

I tried to do that during development of this class, however they are so different that I preferred to keep them apart. 

Why are they different?
- BTreeIndex is a Key-Value index with all keys of the same length and all values of the same length. BTree set has no values and allows variable-length keys.
    - This allows BTreeIndex Nodes/Pages to more efficiently pack the data together (which is necessary (see next bullet)) and execute binary searches for lookups. For BTreeSet I have to keep a header with pointers to data so I can "jump" to an entry.
- BTreeIndex is meant to store data on an append-only medium, while BTreeSet couldn't care less (literally). BtreeIndex is stored directly on Tier2/LTS, while BTreeSet uses the Hash Table Segment itself to store nodes (each node is a Key-Value entry in the Table Segment it keeps keys for). As such, BTreeSet does a lot less work in managing data integrity and compaction (it does none of that!) since it delegates all of it to the Table Segment underneath it.
    - Due to the fact that BTreeIndex stores on Append only media, it has to deal with write amplification so the amount of data that we need to write must be kept to a minimum, hence the branching factor of the node. BTreeIndex has a max node/page size of 32KB, which gives it a branching factor of about 1000 (more or less), thus reducing the depth of the tree and the size of the writes on updates.
    - BTreeSet has no such limitation so its max page/node size is whatever the TableSegment allows as values (1MB as of today). Given that it has variable-length keys, the branching factor will vary, however there is a lower bound on it due to the fact that keys themselves have an upper bound on their length.

Reply:
Ok, I searched and couldn't find an issue, so I created #4840, and assigned it to you for now, feel free to reassign as needed. You can resolve this thread once you see this message, I want to make sure you see the created issue.